### DISCUSSION TITLE
Am I right that queue is implemented as a circular buffer? Why is there a dedicated circular buffer then?

### DISCUSSION THREAD

**Discussion by positron96** [ETLCPP/etl#526] [Q&A] [answered]
Given that etl queue is of fixed size, it might be reasonable to implement it as a circular buffer anyway. A quick glance at the code shows that is uses fixed-size buffer and read/write pointers inside that buffer, so it seems to me that it is a circular buffer under the hood.
As I understand, queue has more usable interfaces for different usages as there are variants for thread-safe usage from ISRs or between threads. 
So I'd assume that `etl::circular_buffer` does roughtly the same. But since it does not have thread-safe variants, what's the point of it? Or are there implementation differences that don't allow the usage of queue as a circular buffer?

**Comment by jwellbelove** [comment, marked as answer]
All of the queues were created long before `etl::circular_buffer` and therefore got multiple variants. There is a  fundamental difference between the queues and the circular buffer in that when a queue becomes full, that's it. No more data can be pushed until an item is pulled.
By contrast, the circular buffer will just over-write old data with new as the input pointer loops round. Circular buffers are usually used in low level interrupt code and protected by interrupt enable/disable code.

**Reply by positron96** [reply]
Huh, I never realized that circular_buffer allows overwriting oldest values. Thank you pointing that out.
